File: leet-code/234. Palindrome Linkded List/PalindromeLinkedList234_2.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;


public class PalindromeLinkedList234_2 {
    public boolean isPalindrome(ListNode head) {
        List<Integer> nums = new ArrayList<>();
        while (head != null) {
            nums.add(head.val);
            head = head.next;
        }
        List<Integer> reversedNums = new ArrayList<>(nums);
        Collections.reverse(reversedNums);
        return nums.equals(reversedNums);
    }
}

File: leet-code/234. Palindrome Linkded List/PalindromeLinkedList234_2.kt
package com.example

class PalindromeLinkedList234_2 {
    fun isPalindrome(head: ListNode?): Boolean {
        var prev: ListNode? = null
        var slow = head
        var fast = head
        while (fast?.next != null) {
            prev = slow
            slow = slow!!.next
            fast = fast.next!!.next
        }
        // 연결 끊기
        prev?.next = null

        prev = null
        while (slow != null) {
            var next = slow.next
            slow.next = prev
            prev = slow
            slow = next
        }
        slow = prev

        var cur = head
        while (cur != null) {
            if (cur.`val` != slow!!.`val`) {
                return false
            }
            cur = cur.next
            slow = slow.next
        }
        return true
    }
}
